### PR TITLE
Resources improvements

### PR DESCRIPTION
**Pull request by cormacpayne**
## Description

Fix for issue https://github.com/Azure/azure-powershell/issues/5903

- Update AD cmdlets so they conform with PowerShell standards
- Add new parameter set to `New-AzureRmADServicePrincipal` for simple creation of a service principal
- Remove `Find-AzureRmResource` and update `Get-AzureRmResource` to work in basic scenarios
- Remove `Find-AzureRmResourceGroup` and move missing functionality into `Get-AzureRmResourceGroup`

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/Common/Commands.Common.Graph.RBAC/ActiveDirectory/ActiveDirectoryClient.cs [hunk 1]
```diff
@@ -765,17 +650,14 @@ public PSADServicePrincipal CreateServicePrincipal(CreatePSServicePrincipalParam
...
-        public PSADServicePrincipal RemoveServicePrincipal(string objectId)
+        public PSADServicePrincipal RemoveServicePrincipal(Guid objectId)
         {
-            ADObjectFilterOptions options = new ADObjectFilterOptions
-            {
-                Id = objectId.ToString()
-            };
-
-            PSADServicePrincipal servicePrincipal = FilterServicePrincipals(options).FirstOrDefault();
+            var objId = objectId.ToString();
```
Comment (markcowl) [inline src/Common/Commands.Common.Graph.RBAC/ActiveDirectory/ActiveDirectoryClient.cs]: better to use a descriptive identifier, like objectIdString

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    using Properties;
+    using Rest.Azure;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+
+    /// <summary>
+    /// This attribute will allow the user to autocomplete the -ResourceType parameter of a cmdlet with valid resource types
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs]: Can we split this into a separate PR?  I want to make sure we get in as much of this as possible

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                        try
+                        {
+                            var instance = AzureSession.Instance;
+                            var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            // Retrieve only the first page of resource types to display to the user
+                            var resourceTypes = client.Providers.ListAsync();
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs]: why only one page?

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                            else
+                            {
+                                throw new InvalidOperationException("client.ResourceGroups call timed out");
+                            }
+#endif
+                        }
+
+                        catch (Exception ex)
+                        {
+                            if (ex == null) { }
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs]: I think this check should also be in the ifdef

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                        var tempScopeList = new List<string>();
+                        try
+                        {
+                            var instance = AzureSession.Instance;
+                            var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                                context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                                instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                                instance.ClientFactory.GetCustomHandlers());
+                            client.SubscriptionId = context.Subscription.Id;
+                            // Retrieve only the first page of ResourceGroups to display to the user
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs]: I wonder if we can't pass in the parameter value through to the resource groups call to make the first page of results more effective

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Generated/SubscriptionsOperationsExtensions.cs [hunk 6]
```diff
@@ -167,9 +168,9 @@ public static IPage<Subscription> ListNext(this ISubscriptionsOperations operati
             /// <returns>
             /// PageEnumerable object containing all of the subscriptions for the given tenant used in the client.
             /// </returns>
-            public static PageEnumerable ListAll(this ISubscriptionClient client)
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/Generated/SubscriptionsOperationsExtensions.cs]: I don't think we can remove these methods - we need to only make addivie changes to the common code

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerable.cs [hunk 7]
```diff
@@ -19,44 +19,31 @@
 using System.Collections.Generic;
 using System.Threading;
 
-namespace Microsoft.Azure.Internal.Subscriptions.Models
+namespace Microsoft.Azure.Commands.ResourceManager.Common.Paging
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerable.cs]: same comment - we cannot make breaking changes here.  If we need to substantially redesign it, we should define a new class.

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs [hunk 8]
```diff
@@ -17,23 +17,36 @@
 using System.Collections.Generic;
 using System.Collections;
 
-namespace Microsoft.Azure.Internal.Subscriptions.Models
```
Comment (markcowl) [inline src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs]: same comment

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/ModelExtensions.cs [hunk 9]
```diff
@@ -113,7 +113,8 @@ public static string GetDisplayNameForADObject(string objectId, ActiveDirectoryC
                     }
                     else if (obj.Type.Equals("serviceprincipal", StringComparison.InvariantCultureIgnoreCase))
                     {
-                        var servicePrincipal = adClient.FilterServicePrincipals(new ADObjectFilterOptions { Id = objectId }).FirstOrDefault();
+                        var odataQuery = new Rest.Azure.OData.ODataQuery<Graph.RBAC.Version1_6.Models.ServicePrincipal>(s => s.ObjectId == objectId);
```
Comment (markcowl) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Models/ModelExtensions.cs]: we should let the keyvault team look at these changes

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/SdkClient/ResourceManagerSdkClient.cs [hunk 10]
```diff
@@ -607,6 +608,28 @@ public virtual List<PSResourceGroup> FilterResourceGroups(string name, Hashtable
                     resourceGroups.AddRange(listResult);
                 }
 
+                if (!string.IsNullOrEmpty(name))
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/SdkClient/ResourceManagerSdkClient.cs]: ToDo:  This would be really nice extracted out into a helper like:

```C#
IEnumerable<T> FilterByWildCard<T>(IEnumerable<T> source, Func<T, string> getKey, string parameterValue)
```

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/SdkModels/Resources/PSResource.cs [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+using Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkExtensions;
+using Microsoft.Azure.Management.ResourceManager.Models;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.SdkModels
+{
+    public class PSResource
+    {
+        public string Id { get; set; }
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/SdkModels/Resources/PSResource.cs]: Need ResourceId for backward-compatible cmdlet piping, right?

**Comment by markcowl** [comment]
@maddieclayton Can we get someone from KeyVaule to pass on the small changes here to KeyVault cmdlets?

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs (hunk 3 above)
Comment (cormacpayne) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ResourceTypeCompleter.cs]: @markcowl looks like this was a copy-paste issue -- I changed this so it retrieves all pages

File: src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs (hunk 4 above)
Comment (cormacpayne) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/ArgumentCompleters/ScopeCompleter.cs]: @markcowl this was modeled after the `ResourceGroupCompleter`, but I believe that it's here so that the release build won't fail due to an unused `ex` variable

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Generated/SubscriptionsOperationsExtensions.cs (hunk 6 above)
Comment (cormacpayne) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/Generated/SubscriptionsOperationsExtensions.cs]: @markcowl reverted `PageEnumerable` and `PageEnumerator` to their original form (only `SubscriptionClient` is handled) and moved all the new functionality into `GenericPageEnumerable` and `GenericPageEnumerator`

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerable.cs (hunk 7 above)
Comment (cormacpayne) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerable.cs]: @markcowl this has been reverted

File: src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs (hunk 8 above)
Comment (cormacpayne) [reply src/ResourceManager/Common/Commands.ResourceManager.Common/Paging/PageEnumerator.cs]: @markcowl this has been reverted

File: src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmADUser.md [hunk 12]
```diff
@@ -81,7 +97,39 @@ Accept pipeline input: False
 Accept wildcard characters: False
 ```
 
+### -First
+Gets the first N objects.
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmADUser.md]: I would say The maximum number of objects to return

File: src/ResourceManager/Resources/Commands.Resources/help/Update-AzureRmADApplication.md [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+### InputObjectWithUpdateParamsParameterSet
+```
+Update-AzureRmADApplication -InputObject <PSADApplication> [-DisplayName <String>] [-HomePage <String>]
+ [-IdentifierUri <String[]>] [-ReplyUrl <String[]>] [-AvailableToOtherTenants <Boolean>]
+ [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [<CommonParameters>]
+```
+
+## DESCRIPTION
+Updates an existing azure active directory application.
+To update the credentials associated with this application, please use New-AzureRmADAppCredential cmdlet.
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/help/Update-AzureRmADApplication.md]: ```<nit>```
.. please use the New-AzureRmADAppCredential...
```</nit>```

File: src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmResource.md [hunk 14]
```diff
@@ -9,95 +9,150 @@ schema: 2.0.0
...
-### GetResourceCollection
+This command gets all of the resources whose resource group they belong in beings with "other".
+
+### Example 4: Get all resources with a given name
+
 ```
-Get-AzureRmResource [-ResourceType <String>] [-ExtensionResourceType <String>] [-ExpandProperties]
- [-IsCollection] [-ODataQuery <String>] [-ResourceGroupName <String>] [-ApiVersion <String>] [-Pre]
- [-DefaultProfile <IAzureContextContainer>] [<CommonParameters>]
+PS C:\> Get-AzureRmResource -Name testVM | ft
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmResource.md]: This output is | fl

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADServicePrincipalCommand.cs [hunk 15]
```diff
@@ -133,53 +216,180 @@ public override void ExecuteCmdlet()
...
+                        Subscription = subscriptionId
+                    },
+                    CanDelegate = false
+                };
+
+                for (var i = 0; i < 6; i++)
+                {
+                    try
+                    {
+                        Thread.Sleep(5000);
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADServicePrincipalCommand.cs]: We should use our own TestMockSupport.Delay method so this doesn't take forever in playback

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/GetAzureADUserCommand.cs [hunk 16]
```diff
@@ -50,22 +50,20 @@ public class GetAzureADUserCommand : ActiveDirectoryBaseCmdlet
...
             ExecutionBlock(() =>
             {
-                do
+                ADObjectFilterOptions options = new ADObjectFilterOptions
                 {
-                    WriteObject(ActiveDirectoryClient.FilterUsers(options), true);
+                    SearchString = SearchString,
+                    UPN = UserPrincipalName,
+                    Id = ObjectId == Guid.Empty ? null : ObjectId.ToString(),
+                    Paging = true,
```
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/GetAzureADUserCommand.cs]: @markcowl it used to signal that the client should follow the `nextPageLink`, but that is now handled by `GenericPageEnumerable`, so this isn't necessarily needed (but I will keep it around 😀 )

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/UpdateAzureADServicePrincipalCommand.cs [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = ParameterSet.InputObjectWithDisplayName, HelpMessage = "The service principal object.")]
+        [ValidateNotNullOrEmpty]
+        public PSADServicePrincipal InputObject { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ParameterSet.SpObjectIdWithDisplayName, HelpMessage = "The display name for the application.")]
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = ParameterSet.SPNWithDisplayName, HelpMessage = "The display name for the application.")]
+        [Parameter(Mandatory = true, ParameterSetName = ParameterSet.InputObjectWithDisplayName, HelpMessage = "The display name for the application")]
+        [ValidateNotNullOrEmpty]
+        public string DisplayName { get; set; }
```
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/UpdateAzureADServicePrincipalCommand.cs]: @markcowl added `Homepage`, `IdentifierUri`, `KeyCredential` and `PasswordCredential` passwords

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/GetAzureADSpCredentialCommand.cs [hunk 18]
```diff
@@ -21,21 +23,32 @@ namespace Microsoft.Azure.Commands.ActiveDirectory
     /// Gets AD service principal credentials.
     /// </summary>
     [Cmdlet(VerbsCommon.Get, "AzureRmADSpCredential", DefaultParameterSetName = ParameterSet.ObjectId), OutputType(typeof(PSADCredential))]
+    [Alias("Get-AzureRmADServicePrincipalCredential")]
```
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/GetAzureADSpCredentialCommand.cs]: @markcowl I think we agreed to skip this comment (I can add the alias if that's not the case)

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADSpCredentialCommand.cs [hunk 19]
```diff
@@ -25,76 +26,91 @@ namespace Microsoft.Azure.Commands.ActiveDirectory
     /// Creates a new AD servicePrincipal Credential.
     /// </summary>
     [Cmdlet(VerbsCommon.New, "AzureRmADSpCredential", DefaultParameterSetName = ParameterSet.SpObjectIdWithPassword, SupportsShouldProcess = true), OutputType(typeof(PSADCredential))]
+    [Alias("New-AzureRmADServicePrincipalCredential")]
```
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/NewAzureADSpCredentialCommand.cs]: @markcowl same comment here about skipping this comment

File: src/ResourceManager/Resources/Commands.Resources/help/Remove-AzureRmADGroupMember.md [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+
+Required: True
+Position: Named
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -MemberUserPrincipalName
+The UPN of the member(s) to remove.```yaml
```
Comment (markcowl) [inline src/ResourceManager/Resources/Commands.Resources/help/Remove-AzureRmADGroupMember.md]: yaml issue

File: src/Common/Commands.Common.Graph.RBAC/ActiveDirectory/ActiveDirectoryClientExtensions.cs [hunk 21]
```diff
@@ -59,7 +59,8 @@ public static PSADObject ToPSADObject(this AADObject obj)
                     DisplayName = obj.DisplayName,
                     Type = obj.ObjectType,
                     Id = new Guid(obj.ObjectId),
-                    SecurityEnabled = obj.SecurityEnabled
+                    SecurityEnabled = obj.SecurityEnabled,
+                    MailNickname = obj.Mail
```
Comment (darshanhs90) [inline src/Common/Commands.Common.Graph.RBAC/ActiveDirectory/ActiveDirectoryClientExtensions.cs]: >MailNickname = obj.Mail [](start = 20, length = 23)

for a group the mail will be always null,do we need this?
[https://msdn.microsoft.com/en-us/library/azure/ad/graph/api/groups-operations](https://msdn.microsoft.com/en-us/library/azure/ad/graph/api/groups-operations)

File: src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/UpdateAzureADServicePrincipalCommand.cs [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+using System.Collections.Generic;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.ActiveDirectory
+{
+    /// <summary>
+    /// Updates an exisitng service principal.
+    /// </summary>
+    [Cmdlet(VerbsData.Update, "AzureRmADServicePrincipal", DefaultParameterSetName = ParameterSet.SpObjectIdWithDisplayName, SupportsShouldProcess = true), OutputType(typeof(PSADServicePrincipal))]
+    [Alias("Set-AzureRmADServicePrincipal")]
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/ActiveDirectory/UpdateAzureADServicePrincipalCommand.cs]: >Set-AzureRmADServicePrincipa [](start = 12, length = 28)

why the change from Set to update?
isnt it confusing/inconsistent wrt other commandlets?
for some we have set and for others we have update

**Comment by markcowl** [comment]
nightly build (Cloudshell prereq) here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-nightly/122/

**Comment by cormacpayne** [comment]
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/781/

File: src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyVaultManagementCmdletBase.cs [hunk 23]
```diff
@@ -265,7 +265,8 @@ private string GetObjectIdBySpn(string spn)
             if (!string.IsNullOrWhiteSpace(spn))
             {
 #if NETSTANDARD
-                var servicePrincipal = ActiveDirectoryClient.FilterServicePrincipals(new ADObjectFilterOptions() { SPN = spn }).SingleOrDefault();
+                var odataQuery = new Rest.Azure.OData.ODataQuery<Graph.RBAC.Version1_6.Models.ServicePrincipal>(s => s.ServicePrincipalNames.Contains(spn));
```
[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/ResourceManager/KeyVault/Commands.KeyVault/Models/KeyVaultManagementCmdletBase.cs]: @RandalliLama the filter is case insensitive

**Comment by cormacpayne** [comment]
CloudShell test: https://azuresdkci.westus2.cloudapp.azure.com/job/ps-cloudshell-test/134/
